apple/internal/aspects/swift_dynamic_framework_aspect.bzl

```diff
@@ -0,0 +1,151 @@
+    doc = """
+Provider that collects artifacts required to build a Swift-based dynamic framework.
+""",
+)
+
+def _swift_target_for_dep(dep):
+    """Returns the target for which the dependency was compiled.
+
+    This is really hacky, but there's no easy way to acquire the Apple CPU for which the target was
+    built. One option would be to let this aspect propagate transitively through deps and have
+    another provider that propagates the CPU, but the model there gets a bit more complicated to
+    follow. With this approach, we avoid propagating the aspect transitively as well.
+
+    This should be cleaned up when b/141931700 is fixed (adding support for ctx.rule.split_attr).
+    """
+    for action in dep.actions:
+        if action.mnemonic == "SwiftCompile":
+            target_found = False
+            for arg in action.argv:
+                if target_found:
+                    return arg
+                if arg == "-target":
+                    target_found = True
+    fail("error: Expected at least one Swift compilation action for target {}.".format(dep.label))
+
+def _swift_arch_for_dep(dep):
+    """Returns the architecture for which the dependency was built."""
+    target = _swift_target_for_dep(dep)
+    return target.split("-", 1)[0]
+
+def _modulemap_contents(module_name):
+    """Returns the contents for the modulemap file for the framework."""
+    return """\
+framework module {module_name} {{
+  header "{module_name}.h"
+  requires objc
+}}
+""".format(module_name = module_name)
+
+def _swift_dynamic_framework_aspect_impl(target, ctx):
+    """Aspect implementation for Swift dynamic framework support."""
+
+    if not hasattr(ctx.rule.attr, "deps"):
+        return []
+    
+    swiftdeps = [x for x in ctx.rule.attr.deps if SwiftInfo in x]
+    ccinfos = [x for x in ctx.rule.attr.deps if CcInfo in x]
+    # If there are no Swift dependencies, return nothing.
+    if not swiftdeps:
+        return []
+
+    if len(swiftdeps) != len(ctx.rule.attr.deps):
+            fail(
+                """\
+error: Found a mix of swift_library and other rule dependencies. Swift dynamic frameworks expect a \
+single swift_library dependency.\
+""",
+            )
+
+    # Collect all relevant artifacts for Swift dynamic framework generation.
+    module_name = None
+    generated_header = None
+    swiftdocs = {}
+    swiftmodules = {}
+    modulemap_file = None
+    for dep in swiftdeps:
+        swiftinfo = dep[SwiftInfo]
+
+        # Use the module_name for the first swiftdep
+        # If there are transitive dependencies, more than one swiftdep will be returned
+        if not module_name:
+            module_name = swiftinfo.module_name
+
+        arch = _swift_arch_for_dep(dep)
+
+        swiftdocs[arch] = swiftinfo.transitive_swiftdocs.to_list().pop()
+        swiftmodules[arch] = swiftinfo.transitive_swiftmodules.to_list().pop()
+        modulemap_file = ctx.actions.declare_file("{}_file.modulemap".format(module_name))
```

Comment:
Given the above statement, "If there are transitive dependencies, more than one swiftdep will be returned", this would declare two files with the same name, right?

Reply:
You are correct. It looks like the aspect gets called twice when the swift_library has dependencies: once for the single swift_library being packaged by the rule and once for the merged target of all the dependencies within that library (if there are any).

I removed the check above module name so it will declare different file names for the dependency targets, but the only one we really care about is the dependency for the single swift_library since that is what gets packaged into the .framework.